pwiz_tools/Shared/CommonUtil/SystemUtil/PathEx.cs

```diff
@@ -255,6 +256,15 @@ public static string EscapePathForXML(string path)
             return path;
         }
 
+        /// <summary>
+        /// Replaces invalid filename characters (\/*:?&lt;>|" and non-printable chars) with a replacement character (default '_').
+        /// </summary>
+        public static string ReplaceInvalidFilenameCharacters(string filename, char replacementChar = '_')
+        {
+            const string illegalFilename = "\\/*:?<>|\"";
+            return new string(filename.Select(t=>t < 0x20 || t == 0x7f || illegalFilename.Contains(t) ? '_' : t).ToArray());

```

Comment:
Maybe use "c" instead of "t" since it is a character. Also, I don't love code that always creates a new array of characters even if nothing is changing. It's a memory thing, and not a speed thing. So, you could use Helpers.AssignIfEquals().

Or you could create a named static function PathEx.IsValidChar(char c) and then do return filename.All(IsValidChar) ? filename : new string(filename.Select(c => IsValidChar(c) ? c : '_').ToArray())

I think the second option would likely result in cleaner code.

Reply:
Although it's probably a bit of premature optimization, I think your second idea is spot on.